File: platform/polyglot/app_yaml_overwrites/py/src/app_yaml_overwrites/compute_registry.py
"""
Compute Registry Module for app_yaml_overwrites package.
Provides registration and execution of compute functions with scope management.
"""

import re
import asyncio
import inspect
from typing import Callable, Dict, Optional, Any, List
from dataclasses import dataclass

from .logger import create as create_logger, ILogger
from .options import ComputeScope
from .errors import ComputeFunctionError, ErrorCode

# Create module-level logger
logger = create_logger("app_yaml_overwrites", "compute_registry.py")

# Type alias for compute functions
ComputeFunction = Callable[..., Any]


@dataclass
class RegisteredFunction:
    """
    Represents a registered compute function with its scope.
    """
    fn: ComputeFunction
    scope: ComputeScope


class ComputeRegistry:
    """
    Registry for compute functions that can be called via {{fn:name}}.

    Features:
    - Function registration with STARTUP or REQUEST scope
    - Name validation (^[a-zA-Z_][a-zA-Z0-9_]*$)
    - STARTUP scope results are cached
    - REQUEST scope functions are executed per-call
    - Supports both sync and async functions
    """

    NAME_PATTERN = re.compile(r'^[a-zA-Z_][a-zA-Z0-9_]*$')

    def __init__(self, logger_instance: Optional[ILogger] = None):
        self._logger = logger_instance or logger
        self._functions: Dict[str, RegisteredFunction] = {}
        self._cache: Dict[str, Any] = {}
        self._logger.debug("ComputeRegistry initialized")

    def register(
        self,
        name: str,
        fn: ComputeFunction,
        scope: ComputeScope = ComputeScope.REQUEST
    ) -> None:
        """
        Register a compute function with the given scope.

        Args:
            name: Function name (must match ^[a-zA-Z_][a-zA-Z0-9_]*$)
            fn: The function to register
            scope: ComputeScope.STARTUP (cached) or ComputeScope.REQUEST (per-call) or ComputeScope.GLOBAL (both, never cached)

        Raises:
            ValueError: If name is invalid
        """
        self._validate_name(name)
        self._logger.debug(f"Registering function: {name} with scope: {scope.value}")
        self._functions[name] = RegisteredFunction(fn=fn, scope=scope)
        self._logger.info(f"Function registered: {name}")

    def unregister(self, name: str) -> bool:
        """
        Unregister a compute function.

        Args:
            name: Function name to unregister

        Returns:
            True if function was unregistered, False if not found
        """
        if name in self._functions:
            self._logger.debug(f"Unregistering function: {name}")
            del self._functions[name]
            # Also remove from cache if present
            self._cache.pop(name, None)
            self._logger.info(f"Function unregistered: {name}")
            return True
        return False

    def has(self, name: str) -> bool:
        """
        Check if a function is registered.

        Args:
            name: Function name to check

        Returns:
            True if function is registered
        """
        return name in self._functions

    def list(self) -> List[str]:
        """
        List all registered function names.

        Returns:
            List of function names
        """
        return list(self._functions.keys())

    def get_scope(self, name: str) -> Optional[ComputeScope]:
        """
        Get the scope of a registered function.

        Args:
            name: Function name

        Returns:
            ComputeScope if found, None otherwise
        """
        if name in self._functions:
            return self._functions[name].scope
        return None

    def clear(self) -> None:
        """
        Clear all registered functions and cache.
        """
        self._logger.debug("Clearing registry")
        self._functions.clear()
        self._cache.clear()

    def clear_cache(self) -> None:
        """
        Clear only the result cache (keep registrations).
        """
        self._logger.debug("Clearing result cache")
        self._cache.clear()

    async def resolve(
        self,
        name: str,
        context: Optional[Dict[str, Any]] = None,
        property_path: Optional[str] = None
    ) -> Any:
        """
        Resolve (execute) a compute function.

        Args:
            name: Function name to execute
            context: Optional context dict passed to the function
            property_path: The path to the property being computed (e.g., "providers.gemini_openai.api_key")

        Returns:
            The function's return value

        Raises:
            ComputeFunctionError: If function not found or execution fails
        """
        self._logger.debug(f"Resolving function: {name}", property_path=property_path)

        if name not in self._functions:
            self._logger.warn(f"Compute function not found: {name}")
            raise ComputeFunctionError(
                f"Compute function not found: {name}",
                ErrorCode.COMPUTE_FUNCTION_NOT_FOUND,
                {"name": name}
            )

        reg_fn = self._functions[name]

        # Check cache for STARTUP functions (cache key includes property_path for uniqueness)
        cache_key = f"{name}:{property_path}" if property_path else name
        if reg_fn.scope == ComputeScope.STARTUP and cache_key in self._cache:
            self._logger.debug(f"Returning cached value for: {cache_key}")
            return self._cache[cache_key]

        try:
            result = await self._execute_function(reg_fn.fn, context, property_path)

            # Cache result if STARTUP scope
            if reg_fn.scope == ComputeScope.STARTUP:
                self._cache[cache_key] = result
                self._logger.debug(f"Cached result for STARTUP function: {cache_key}")

            return result

        except Exception as e:
            self._logger.error(f"Function execution failed: {name}, error: {str(e)}")
            raise ComputeFunctionError(
                f"Compute function failed: {name}",
                ErrorCode.COMPUTE_FUNCTION_FAILED,
                {"name": name, "original_error": str(e)}
            ) from e

    def _get_required_param_count(self, fn: ComputeFunction) -> int:
        """
        Get the number of required parameters for a function using inspect.

        Returns:
            Number of required parameters (parameters without defaults)
        """
        try:
            sig = inspect.signature(fn)
            required = 0
            for param in sig.parameters.values():
                # Count parameters that don't have defaults and aren't *args/**kwargs
                if param.default is inspect.Parameter.empty and \
                   param.kind not in (inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD):
                    required += 1
            return required
        except (ValueError, TypeError):
            # If we can't inspect, assume it takes context
            return 1

    async def _execute_function(
        self,
        fn: ComputeFunction,
        context: Optional[Dict[str, Any]],
        property_path: Optional[str] = None
    ) -> Any:
        """
        Execute a function, handling both sync and async functions.
        Uses inspect to determine correct function signature instead of catching TypeError.

        Function signatures supported:
        - fn(ctx) - Legacy: context only
        - fn(ctx, property_path) - New: context and property path
        - fn() - No arguments (fallback)

        Args:
            fn: The compute function to execute
            context: Context dict with env, config, request, etc.
            property_path: Path to the property being computed (e.g., "providers.gemini_openai.api_key")
        """
        # Use inspect to determine parameter count - avoids catching internal TypeErrors
        param_count = self._get_required_param_count(fn)

        if asyncio.iscoroutinefunction(fn):
            if param_count >= 2:
                return await fn(context, property_path)
            elif param_count == 1:
                return await fn(context)
            else:
                return await fn()
        else:
            if param_count >= 2:
                result = fn(context, property_path)
            elif param_count == 1:
                result = fn(context)
            else:
                result = fn()

            if asyncio.iscoroutine(result):
                return await result
            return result

    def _validate_name(self, name: str) -> None:
        """
        Validate a function name.

        Args:
            name: Function name to validate

        Raises:
            ValueError: If name is invalid
        """
        if not name:
            raise ValueError("Function name cannot be empty")
        if not self.NAME_PATTERN.match(name):
            raise ValueError(
                f"Invalid function name: {name}. "
                f"Must match pattern: ^[a-zA-Z_][a-zA-Z0-9_]*$"
            )


def create_registry(logger_instance: Optional[ILogger] = None) -> ComputeRegistry:
    """
    Factory function to create a ComputeRegistry instance.

    Args:
        logger_instance: Optional custom logger

    Returns:
        ComputeRegistry instance
    """
    return ComputeRegistry(logger_instance)

File: platform/polyglot/app_yaml_overwrites/mjs/src/compute-registry.ts
/**
 * Compute Registry Module for app-yaml-overwrites package.
 * Provides registration and execution of compute functions with scope management.
 */

import { create as createLogger, type ILogger } from './logger.js';
import { ComputeScope } from './options.js';
import { ComputeFunctionError, ErrorCode } from './errors.js';

// Create module-level logger
const logger = createLogger('app-yaml-overwrites', 'compute-registry.ts');

/**
 * Type for compute functions - can be sync or async, with optional context.
 */
export type ComputeFunction = (context?: any) => any | Promise<any>;

/**
 * Represents a registered compute function with its scope.
 */
export interface RegisteredFunction {
    fn: ComputeFunction;
    scope: ComputeScope;
}

/**
 * Registry for compute functions that can be called via {{fn:name}}.
 *
 * Features:
 * - Function registration with STARTUP or REQUEST scope
 * - Name validation (^[a-zA-Z_][a-zA-Z0-9_]*$)
 * - STARTUP scope results are cached
 * - REQUEST scope functions are executed per-call
 * - Supports both sync and async functions
 */
export class ComputeRegistry {
    private logger: ILogger;
    private functions: Map<string, RegisteredFunction>;
    private cache: Map<string, any>;

    private static readonly NAME_PATTERN = /^[a-zA-Z_][a-zA-Z0-9_]*$/;

    constructor(loggerInstance?: ILogger) {
        this.logger = loggerInstance || logger;
        this.functions = new Map();
        this.cache = new Map();
        this.logger.debug('ComputeRegistry initialized');
    }

    /**
     * Register a compute function with the given scope.
     *
     * @param name - Function name (must match ^[a-zA-Z_][a-zA-Z0-9_]*$)
     * @param fn - The function to register
     * @param scope - ComputeScope.STARTUP (cached) or ComputeScope.REQUEST (per-call)
     * @throws Error if name is invalid
     */
    public register(
        name: string,
        fn: ComputeFunction,
        scope: ComputeScope = ComputeScope.REQUEST
    ): void {
        this.validateName(name);
        this.logger.debug(`Registering function: ${name} with scope: ${scope}`);
        this.functions.set(name, { fn, scope });
        this.logger.info(`Function registered: ${name}`);
    }

    /**
     * Unregister a compute function.
     *
     * @param name - Function name to unregister
     * @returns True if function was unregistered, false if not found
     */
    public unregister(name: string): boolean {
        if (this.functions.has(name)) {
            this.logger.debug(`Unregistering function: ${name}`);
            this.functions.delete(name);
            // Also remove from cache if present
            this.cache.delete(name);
            this.logger.info(`Function unregistered: ${name}`);
            return true;
        }
        return false;
    }

    /**
     * Check if a function is registered.
     *
     * @param name - Function name to check
     * @returns True if function is registered
     */
    public has(name: string): boolean {
        return this.functions.has(name);
    }

    /**
     * List all registered function names.
     *
     * @returns Array of function names
     */
    public list(): string[] {
        return Array.from(this.functions.keys());
    }

    /**
     * Get the scope of a registered function.
     *
     * @param name - Function name
     * @returns ComputeScope if found, undefined otherwise
     */
    public getScope(name: string): ComputeScope | undefined {
        return this.functions.get(name)?.scope;
    }

    /**
     * Clear all registered functions and cache.
     */
    public clear(): void {
        this.logger.debug('Clearing registry');
        this.functions.clear();
        this.cache.clear();
    }

    /**
     * Clear only the result cache (keep registrations).
     */
    public clearCache(): void {
        this.logger.debug('Clearing result cache');
        this.cache.clear();
    }

    /**
     * Resolve (execute) a compute function.
     *
     * @param name - Function name to execute
     * @param context - Optional context object passed to the function
     * @param propertyPath - The path to the property being computed (e.g., "providers.gemini_openai.api_key")
     * @returns The function's return value
     * @throws ComputeFunctionError if function not found or execution fails
     */
    public async resolve(name: string, context?: any, propertyPath?: string): Promise<any> {
        this.logger.debug(`Resolving function: ${name}`, { propertyPath });

        const regFn = this.functions.get(name);
        if (!regFn) {
            this.logger.warn(`Compute function not found: ${name}`);
            throw new ComputeFunctionError(
                `Compute function not found: ${name}`,
                ErrorCode.COMPUTE_FUNCTION_NOT_FOUND,
                { name }
            );
        }

        // Check cache for STARTUP functions (cache key includes propertyPath for uniqueness)
        const cacheKey = propertyPath ? `${name}:${propertyPath}` : name;
        if (regFn.scope === ComputeScope.STARTUP && this.cache.has(cacheKey)) {
            this.logger.debug(`Returning cached value for: ${cacheKey}`);
            return this.cache.get(cacheKey);
        }

        try {
            const result = await this.executeFunction(regFn.fn, context, propertyPath);

            // Cache result if STARTUP scope
            if (regFn.scope === ComputeScope.STARTUP) {
                this.cache.set(cacheKey, result);
                this.logger.debug(`Cached result for STARTUP function: ${cacheKey}`);
            }

            return result;
        } catch (e: any) {
            this.logger.error(`Function execution failed: ${name}, error: ${e.message}`);
            throw new ComputeFunctionError(
                `Compute function failed: ${name}`,
                ErrorCode.COMPUTE_FUNCTION_FAILED,
                { name, originalError: e.message }
            );
        }
    }

    /**
     * Execute a function, handling both sync and async functions.
     * Passes context and propertyPath to the function.
     *
     * Function signatures supported:
     * - fn(ctx, propertyPath) - New: context and property path
     * - fn(ctx) - Legacy: context only
     * - fn() - No arguments (fallback)
     *
     * @param fn - The compute function to execute
     * @param context - Context object with env, config, request, etc.
     * @param propertyPath - Path to the property being computed (e.g., "providers.gemini_openai.api_key")
     */
    private async executeFunction(fn: ComputeFunction, context?: any, propertyPath?: string): Promise<any> {
        // Try with both context and propertyPath first
        try {
            const result = (fn as any)(context, propertyPath);
            return result instanceof Promise ? await result : result;
        } catch (e: any) {
            // If TypeError, try calling with context only
            if (e instanceof TypeError) {
                try {
                    const result = fn(context);
                    return result instanceof Promise ? await result : result;
                } catch (e2: any) {
                    // If still TypeError, try calling with no-args
                    if (e2 instanceof TypeError) {
                        const result = (fn as any)();
                        return result instanceof Promise ? await result : result;
                    }
                    throw e2;
                }
            }
            throw e;
        }
    }

    /**
     * Validate a function name.
     *
     * @param name - Function name to validate
     * @throws Error if name is invalid
     */
    private validateName(name: string): void {
        if (!name) {
            throw new Error('Function name cannot be empty');
        }
        if (!ComputeRegistry.NAME_PATTERN.test(name)) {
            throw new Error(
                `Invalid function name: ${name}. ` +
                `Must match pattern: ^[a-zA-Z_][a-zA-Z0-9_]*$`
            );
        }
    }
}

/**
 * Factory function to create a ComputeRegistry instance.
 *
 * @param loggerInstance - Optional custom logger
 * @returns ComputeRegistry instance
 */
export function createRegistry(loggerInstance?: ILogger): ComputeRegistry {
    return new ComputeRegistry(loggerInstance);
}
